File: src/shared/states.go
package shared

import (
	"ae_invoice/src/logger"
	"sort"
)

type StateInfo struct {
	MinCode uint32
	MaxCode uint32
	Cities  []string
}

var States = map[string]StateInfo{
	"Andaman and Nicobar Islands": {
		MinCode: 744101, MaxCode: 744304,
		Cities: []string{"Port Blair", "Car Nicobar", "Diglipur", "Rangat", "Mayabunder"},
	},
	"Andhra Pradesh": {
		MinCode: 507130, MaxCode: 535594,
		Cities: []string{"Visakhapatnam", "Vijayawada", "Guntur", "Tirupati", "Kurnool", "Kakinada", "Nellore", "Rajahmundry"},
	},
	"Arunachal Pradesh": {
		MinCode: 790001, MaxCode: 792131,
		Cities: []string{"Itanagar", "Naharlagun", "Pasighat", "Tezpur", "Bomdila", "Ziro"},
	},
	"Assam": {
		MinCode: 781001, MaxCode: 788931,
		Cities: []string{"Guwahati", "Silchar", "Dibrugarh", "Jorhat", "Nagaon", "Tinsukia", "Tezpur", "Bongaigaon"},
	},
	"Bihar": {
		MinCode: 800001, MaxCode: 855117,
		Cities: []string{"Patna", "Gaya", "Muzaffarpur", "Bhagalpur", "Darbhanga", "Purnia", "Arrah", "Begusarai"},
	},
	"Chandigarh": {
		MinCode: 140119, MaxCode: 160102,
		Cities: []string{"Chandigarh"},
	},
	"Chhattisgarh": {
		MinCode: 490001, MaxCode: 497778,
		Cities: []string{"Raipur", "Bhilai", "Bilaspur", "Korba", "Durg", "Rajnandgaon", "Jagdalpur"},
	},
	"Dadra and Nagar Haveli and Daman and Diu": {
		MinCode: 362520, MaxCode: 396240,
		Cities: []string{"Daman", "Diu", "Silvassa"},
	},
	"Delhi": {
		MinCode: 110001, MaxCode: 110097,
		Cities: []string{"New Delhi", "Delhi", "Dwarka", "Rohini", "Shahdara", "Janakpuri", "Pitampura"},
	},
	"Goa": {
		MinCode: 403001, MaxCode: 403806,
		Cities: []string{"Panaji", "Margao", "Vasco da Gama", "Mapusa", "Ponda", "Calangute"},
	},
	"Gujarat": {
		MinCode: 360001, MaxCode: 396590,
		Cities: []string{"Ahmedabad", "Surat", "Vadodara", "Rajkot", "Bhavnagar", "Jamnagar", "Gandhinagar", "Anand"},
	},
	"Haryana": {
		MinCode: 121001, MaxCode: 136156,
		Cities: []string{"Faridabad", "Gurugram", "Panipat", "Ambala", "Yamunanagar", "Rohtak", "Hisar", "Karnal"},
	},
	"Himachal Pradesh": {
		MinCode: 171001, MaxCode: 177601,
		Cities: []string{"Shimla", "Dharamshala", "Solan", "Mandi", "Kullu", "Hamirpur", "Una"},
	},
	"Jammu and Kashmir (including Ladakh)": {
		MinCode: 180001, MaxCode: 194404,
		Cities: []string{"Srinagar", "Jammu", "Leh", "Anantnag", "Baramulla", "Kargil", "Sopore"},
	},
	"Jharkhand": {
		MinCode: 813208, MaxCode: 835325,
		Cities: []string{"Ranchi", "Jamshedpur", "Dhanbad", "Bokaro", "Hazaribagh", "Deoghar", "Giridih"},
	},
	"Karnataka": {
		MinCode: 560001, MaxCode: 591346,
		Cities: []string{"Bengaluru", "Mysuru", "Hubli", "Mangaluru", "Belagavi", "Davanagere", "Ballari", "Tumkur"},
	},
	"Kerala": {
		MinCode: 670001, MaxCode: 695615,
		Cities: []string{"Thiruvananthapuram", "Kochi", "Kozhikode", "Thrissur", "Kollam", "Palakkad", "Alappuzha", "Kannur"},
	},
	"Lakshadweep": {
		MinCode: 682551, MaxCode: 682559,
		Cities: []string{"Kavaratti", "Agatti", "Minicoy", "Amini"},
	},
	"Madhya Pradesh": {
		MinCode: 450001, MaxCode: 488448,
		Cities: []string{"Bhopal", "Indore", "Gwalior", "Jabalpur", "Ujjain", "Sagar", "Dewas", "Satna"},
	},
	"Maharashtra": {
		MinCode: 400001, MaxCode: 445402,
		Cities: []string{"Mumbai", "Pune", "Nagpur", "Nashik", "Aurangabad", "Solapur", "Thane", "Kolhapur", "Amravati"},
	},
	"Manipur": {
		MinCode: 795001, MaxCode: 795159,
		Cities: []string{"Imphal", "Thoubal", "Bishnupur", "Churachandpur", "Senapati"},
	},
	"Meghalaya": {
		MinCode: 783123, MaxCode: 794115,
		Cities: []string{"Shillong", "Tura", "Jowai", "Nongpoh", "Baghmara"},
	},
	"Mizoram": {
		MinCode: 796001, MaxCode: 796901,
		Cities: []string{"Aizawl", "Lunglei", "Champhai", "Serchhip", "Kolasib"},
	},
	"Nagaland": {
		MinCode: 797001, MaxCode: 798627,
		Cities: []string{"Kohima", "Dimapur", "Mokokchung", "Tuensang", "Wokha", "Zunheboto"},
	},
	"Odisha": {
		MinCode: 751001, MaxCode: 770076,
		Cities: []string{"Bhubaneswar", "Cuttack", "Rourkela", "Berhampur", "Sambalpur", "Puri", "Balasore"},
	},
	"Puducherry": {
		MinCode: 533464, MaxCode: 673310,
		Cities: []string{"Puducherry", "Karaikal", "Mahe", "Yanam"},
	},
	"Punjab": {
		MinCode: 140001, MaxCode: 160104,
		Cities: []string{"Ludhiana", "Amritsar", "Jalandhar", "Patiala", "Bathinda", "Mohali", "Pathankot", "Hoshiarpur"},
	},
	"Rajasthan": {
		MinCode: 301001, MaxCode: 345034,
		Cities: []string{"Jaipur", "Jodhpur", "Kota", "Bikaner", "Ajmer", "Udaipur", "Bhilwara", "Alwar"},
	},
	"Sikkim": {
		MinCode: 737101, MaxCode: 737139,
		Cities: []string{"Gangtok", "Namchi", "Geyzing", "Mangan"},
	},
	"Tamil Nadu": {
		MinCode: 600001, MaxCode: 643253,
		Cities: []string{"Chennai", "Coimbatore", "Madurai", "Tiruchirappalli", "Salem", "Tirunelveli", "Vellore", "Erode"},
	},
	"Telangana": {
		MinCode: 500001, MaxCode: 509412,
		Cities: []string{"Hyderabad", "Warangal", "Nizamabad", "Karimnagar", "Khammam", "Ramagundam", "Mahbubnagar"},
	},
	"Tripura": {
		MinCode: 799001, MaxCode: 799290,
		Cities: []string{"Agartala", "Dharmanagar", "Udaipur", "Kailashahar", "Belonia"},
	},
	"Uttar Pradesh": {
		MinCode: 201001, MaxCode: 285223,
		Cities: []string{"Lucknow", "Kanpur", "Varanasi", "Agra", "Prayagraj", "Meerut", "Ghaziabad", "Noida", "Bareilly"},
	},
	"Uttarakhand": {
		MinCode: 244712, MaxCode: 263680,
		Cities: []string{"Dehradun", "Haridwar", "Roorkee", "Haldwani", "Rudrapur", "Rishikesh", "Kashipur"},
	},
	"West Bengal": {
		MinCode: 700001, MaxCode: 743711,
		Cities: []string{"Kolkata", "Howrah", "Durgapur", "Asansol", "Siliguri", "Bardhaman", "Malda", "Kharagpur"},
	},
}

// StateNames is a sorted slice for use in dropdowns
var StateNames []string

func init() {
	// comptime-equivalent length check
	if len(States) != 35 {
		logger.Logger.Fatalf("IndianStates map has %d entries, expected 35", len(States))
	}
	for name := range States {
		StateNames = append(StateNames, name)
	}
	sort.Strings(StateNames)
}

File: src/states.rs
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct StateInfo {
    pub min_code: u32,
    pub max_code: u32,
    pub cities: Vec<&'static str>,
}

/// Returns the full states map. Equivalent to util.States in Go.
pub fn states() -> HashMap<&'static str, StateInfo> {
    let mut m = HashMap::new();
    m.insert("Andaman and Nicobar Islands", StateInfo { min_code: 744101, max_code: 744304, cities: vec!["Port Blair", "Car Nicobar", "Diglipur", "Rangat", "Mayabunder"] });
    m.insert("Andhra Pradesh", StateInfo { min_code: 507130, max_code: 535594, cities: vec!["Visakhapatnam", "Vijayawada", "Guntur", "Tirupati", "Kurnool", "Kakinada", "Nellore", "Rajahmundry"] });
    m.insert("Arunachal Pradesh", StateInfo { min_code: 790001, max_code: 792131, cities: vec!["Itanagar", "Naharlagun", "Pasighat", "Tezpur", "Bomdila", "Ziro"] });
    m.insert("Assam", StateInfo { min_code: 781001, max_code: 788931, cities: vec!["Guwahati", "Silchar", "Dibrugarh", "Jorhat", "Nagaon", "Tinsukia", "Tezpur", "Bongaigaon"] });
    m.insert("Bihar", StateInfo { min_code: 800001, max_code: 855117, cities: vec!["Patna", "Gaya", "Muzaffarpur", "Bhagalpur", "Darbhanga", "Purnia", "Arrah", "Begusarai"] });
    m.insert("Chandigarh", StateInfo { min_code: 140119, max_code: 160102, cities: vec!["Chandigarh"] });
    m.insert("Chhattisgarh", StateInfo { min_code: 490001, max_code: 497778, cities: vec!["Raipur", "Bhilai", "Bilaspur", "Korba", "Durg", "Rajnandgaon", "Jagdalpur"] });
    m.insert("Dadra and Nagar Haveli and Daman and Diu", StateInfo { min_code: 362520, max_code: 396240, cities: vec!["Daman", "Diu", "Silvassa"] });
    m.insert("Delhi", StateInfo { min_code: 110001, max_code: 110097, cities: vec!["New Delhi", "Delhi", "Dwarka", "Rohini", "Shahdara", "Janakpuri", "Pitampura"] });
    m.insert("Goa", StateInfo { min_code: 403001, max_code: 403806, cities: vec!["Panaji", "Margao", "Vasco da Gama", "Mapusa", "Ponda", "Calangute"] });
    m.insert("Gujarat", StateInfo { min_code: 360001, max_code: 396590, cities: vec!["Ahmedabad", "Surat", "Vadodara", "Rajkot", "Bhavnagar", "Jamnagar", "Gandhinagar", "Anand"] });
    m.insert("Haryana", StateInfo { min_code: 121001, max_code: 136156, cities: vec!["Faridabad", "Gurugram", "Panipat", "Ambala", "Yamunanagar", "Rohtak", "Hisar", "Karnal"] });
    m.insert("Himachal Pradesh", StateInfo { min_code: 171001, max_code: 177601, cities: vec!["Shimla", "Dharamshala", "Solan", "Mandi", "Kullu", "Hamirpur", "Una"] });
    m.insert("Jammu and Kashmir (including Ladakh)", StateInfo { min_code: 180001, max_code: 194404, cities: vec!["Srinagar", "Jammu", "Leh", "Anantnag", "Baramulla", "Kargil", "Sopore"] });
    m.insert("Jharkhand", StateInfo { min_code: 813208, max_code: 835325, cities: vec!["Ranchi", "Jamshedpur", "Dhanbad", "Bokaro", "Hazaribagh", "Deoghar", "Giridih"] });
    m.insert("Karnataka", StateInfo { min_code: 560001, max_code: 591346, cities: vec!["Bengaluru", "Mysuru", "Hubli", "Mangaluru", "Belagavi", "Davanagere", "Ballari", "Tumkur"] });
    m.insert("Kerala", StateInfo { min_code: 670001, max_code: 695615, cities: vec!["Thiruvananthapuram", "Kochi", "Kozhikode", "Thrissur", "Kollam", "Palakkad", "Alappuzha", "Kannur"] });
    m.insert("Lakshadweep", StateInfo { min_code: 682551, max_code: 682559, cities: vec!["Kavaratti", "Agatti", "Minicoy", "Amini"] });
    m.insert("Madhya Pradesh", StateInfo { min_code: 450001, max_code: 488448, cities: vec!["Bhopal", "Indore", "Gwalior", "Jabalpur", "Ujjain", "Sagar", "Dewas", "Satna"] });
    m.insert("Maharashtra", StateInfo { min_code: 400001, max_code: 445402, cities: vec!["Mumbai", "Pune", "Nagpur", "Nashik", "Aurangabad", "Solapur", "Thane", "Kolhapur", "Amravati"] });
    m.insert("Manipur", StateInfo { min_code: 795001, max_code: 795159, cities: vec!["Imphal", "Thoubal", "Bishnupur", "Churachandpur", "Senapati"] });
    m.insert("Meghalaya", StateInfo { min_code: 783123, max_code: 794115, cities: vec!["Shillong", "Tura", "Jowai", "Nongpoh", "Baghmara"] });
    m.insert("Mizoram", StateInfo { min_code: 796001, max_code: 796901, cities: vec!["Aizawl", "Lunglei", "Champhai", "Serchhip", "Kolasib"] });
    m.insert("Nagaland", StateInfo { min_code: 797001, max_code: 798627, cities: vec!["Kohima", "Dimapur", "Mokokchung", "Tuensang", "Wokha", "Zunheboto"] });
    m.insert("Odisha", StateInfo { min_code: 751001, max_code: 770076, cities: vec!["Bhubaneswar", "Cuttack", "Rourkela", "Berhampur", "Sambalpur", "Puri", "Balasore"] });
    m.insert("Puducherry", StateInfo { min_code: 533464, max_code: 673310, cities: vec!["Puducherry", "Karaikal", "Mahe", "Yanam"] });
    m.insert("Punjab", StateInfo { min_code: 140001, max_code: 160104, cities: vec!["Ludhiana", "Amritsar", "Jalandhar", "Patiala", "Bathinda", "Mohali", "Pathankot", "Hoshiarpur"] });
    m.insert("Rajasthan", StateInfo { min_code: 301001, max_code: 345034, cities: vec!["Jaipur", "Jodhpur", "Kota", "Bikaner", "Ajmer", "Udaipur", "Bhilwara", "Alwar"] });
    m.insert("Sikkim", StateInfo { min_code: 737101, max_code: 737139, cities: vec!["Gangtok", "Namchi", "Geyzing", "Mangan"] });
    m.insert("Tamil Nadu", StateInfo { min_code: 600001, max_code: 643253, cities: vec!["Chennai", "Coimbatore", "Madurai", "Tiruchirappalli", "Salem", "Tirunelveli", "Vellore", "Erode"] });
    m.insert("Telangana", StateInfo { min_code: 500001, max_code: 509412, cities: vec!["Hyderabad", "Warangal", "Nizamabad", "Karimnagar", "Khammam", "Ramagundam", "Mahbubnagar"] });
    m.insert("Tripura", StateInfo { min_code: 799001, max_code: 799290, cities: vec!["Agartala", "Dharmanagar", "Udaipur", "Kailashahar", "Belonia"] });
    m.insert("Uttar Pradesh", StateInfo { min_code: 201001, max_code: 285223, cities: vec!["Lucknow", "Kanpur", "Varanasi", "Agra", "Prayagraj", "Meerut", "Ghaziabad", "Noida", "Bareilly"] });
    m.insert("Uttarakhand", StateInfo { min_code: 244712, max_code: 263680, cities: vec!["Dehradun", "Haridwar", "Roorkee", "Haldwani", "Rudrapur", "Rishikesh", "Kashipur"] });
    m.insert("West Bengal", StateInfo { min_code: 700001, max_code: 743711, cities: vec!["Kolkata", "Howrah", "Durgapur", "Asansol", "Siliguri", "Bardhaman", "Malda", "Kharagpur"] });

    debug_assert_eq!(m.len(), 35, "States map has {} entries, expected 35", m.len());
    m
}

/// Sorted state names for dropdowns — equivalent to util.StateNames in Go.
pub fn state_names() -> Vec<&'static str> {
    let mut names: Vec<&'static str> = states().into_keys().collect();
    names.sort_unstable();
    names
}
